Add cmake option to control manylinux1 (PEP 513) support.

silee2 opened the pull request:
Add a new CMake option (NGRAPH_MANYLINUX_ENABLE) for Linux to build nGraph in a way that satisfies PEP-513 (manylinux1)
With this change, a single python wheel (per python version) can be used on most modern linux distribution. (Tested on various versions of CentOS and Ubuntu)
The new option currently supports Interpreter and CPU backend (Both DEX and Codegen) only.

Notable changes related to manylinux1
- OpenMP runtime is built from LLVM's OpenMP runtime source to satisfy PEP-513 (w/ NGRAPH_MANYLINUX_ENABLE)
- workaround for aligned_alloc and tmp_filename (w/ NGRAPH_MANYLINUX_ENABLE)
- Added script to create python 2.7, 3.4, 3.5, 3.6, 3.7 `manylinux1_x86_64` wheels from official CentOS 5 docker image for PEP-513
- setup.in.py overrides bdist_wheel.get_tag() and sets `manylinux1_x86_64` if NGRAPH_MANYLINUX_ENABLE
- Added script to build and use clang 7 and clang 8 for manylinux1 build. manylinux1 can target latest marchs

Improvements
- python wheel is built by setup.in.py and tox.in.ini cmake template (legacy setup.py and tox.ini is still kept and up to date to support current CI and ONNX development environment)
- Parallel compilation for setup.in.py
- setup.in.py places shared libraries in the same directory as `_pyngraph` extension.
- setup.in.py collects source files, package_dir and packages automatically.
- Fix incorrect header_resource.hpp
- Make system header file search for Codegen more robust (Redhat/CentOS(w, w/o devtoolset), Ubuntu)
- Download LLVM source instead of cloning repo
- Build LLVM without terminfo and zlib
- Place all shared libraries in NGRAPH_BUILD_DIR after building them and before setting linker search path

To be handled in separate PR
- Unify legacy setup.py, tox.ini and new cmake templates

silee2 commented on contrib/docker/make-manylinux1.sh:
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
```
manylinux1 is only for Linux. Need a separate script to build windows Wheel

rkimballn1 commented on cmake/external_mkldnn.cmake:
```diff
@@ -90,42 +114,84 @@ if(NGRAPH_LIB_VERSIONING_ENABLE)
-    INSTALL_DIR "${EXTERNAL_PROJECTS_ROOT}/mkldnn"
-    EXCLUDE_FROM_ALL TRUE
-    )
+# The 'BUILD_BYPRODUCTS' argument was introduced in CMake 3.2.
```
We don't support ninja so you can remove the two cases here and use only one. We don't need BUILD_BYPRODUCTS so you can remove the else part of this.

rkimballn1 commented on CMakeLists.txt:
```diff
@@ -90,6 +90,59 @@ option(NGRAPH_DEX_ONLY "Build CPU DEX without codegen" FALSE)
+        endif()
+    endif()
+    if(NGRAPH_GPU_ENABLE)
+        message(WARNING "manylinux Python API does not support GPU backend!")
```
All of these should be fatal errors. If the user set bad settings them tell them, don't fix them.

silee2 commented on CMakeLists.txt:
```diff
@@ -106,6 +106,21 @@ option(NGRAPH_DEX_ONLY "Build CPU DEX without codegen" FALSE)
+    endif()
+endif()
+if (NGRAPH_MANYLINUX_ENABLE AND NOT NGRAPH_DEX_ONLY)
+    message(FATAL_ERROR "Please enable option NGRAPH_DEX_ONLY for manylinux Python API!")
```
Added comment.

silee2 commented:
@rkimballn1 @postrational @diyessi @crlishka @aslepko I complete work and removed WIP tag. All CI passes now. You can start the review. I've updated description and listed the changes made to help review.

silee2 commented:
@rkimballn1 I've addressed all of your previous requests. Codegen is now supported.

silee2 commented:
Closing this PR as I have already started breaking down to smaller PRs.

silee2 commented:
Don't delete the branch as I still need it at a reference point.
